src/db-backup.ts

```diff
@@ -0,0 +1,115 @@
+import fs from 'fs'
+import config from 'config'
+import path from 'path'
+import BigNumber from 'bignumber.js'
+import { BlockHeader, Eth } from 'web3-eth'
+
+import { AutoStartStopEventEmitter, NEW_BLOCK_EVENT_NAME } from './blockchain/new-block-emitters'
+import { DbBackUpConfig } from './definitions'
+import { getNewBlockEmitter } from './blockchain/utils'
+
+export type BackUpEntry = { name: string, block: { hash: string, number: BigNumber } }
+
+export function parseBackUps (backUpName: string): BackUpEntry {
+  const [block] = backUpName.split('.')[0].split('-')
+  const [hash, blockNumber] = block.split(':')
+
+  return {
+    name: backUpName,
+    block: { number: new BigNumber(blockNumber), hash }
+  }
+}
+
+export function getBackUps (): BackUpEntry[] {
+  const backupConfig = config.get<DbBackUpConfig>('dbBackUp')
+
+  const backups = fs.readdirSync(path.resolve(backupConfig.path))
+
+  if (backups.length) {
+    return backups
+      .map(parseBackUps)
+      .sort(
+        (a: Record<string, any>, b: Record<string, any>) =>
+          a.block.number.gt(b.block.number) ? -1 : 1
+      )
+  }
+
+  return []
+}
+
+export class DbBackUpJob {
+  readonly newBlockEmitter: AutoStartStopEventEmitter
+  readonly db: string
+  readonly eth: Eth
+  readonly backUpConfig: DbBackUpConfig
+
+  constructor (eth: Eth) {
+    if (!config.has('dbBackUp')) {
+      throw new Error('DB Backup config not exist')
+    }
+    this.backUpConfig = config.get<DbBackUpConfig>('dbBackUp')
+    this.db = config.get<string>('db')
+
+    if (!fs.existsSync(path.resolve(this.backUpConfig.path))) {
+      fs.mkdirSync(path.resolve(this.backUpConfig.path))
+    }
+
+    this.eth = eth
+    this.newBlockEmitter = getNewBlockEmitter(eth)
+  }
+
+  /**
+   * Back-up database if blocks condition met
+   * @return {Promise<void>}
+   * @param block
+   */
+  private async backupDb (block: BlockHeader): Promise<void> {
+    const [lastBackUp, previousBackUp] = getBackUps()
+
+    if (!lastBackUp || new BigNumber(block.number).minus(this.backUpConfig.blocks).gte(lastBackUp.block.number)) {
+      // copy and rename current db
+      await fs.promises.copyFile(this.db, path.resolve(this.backUpConfig.path, `${block.hash}:${block.number}-${this.db}`))
+
+      // remove the oldest version
+      if (previousBackUp) {
+        await fs.promises.unlink(path.resolve(this.backUpConfig.path, previousBackUp.name))
+      }
+    }
+  }
+
+  /**
+   * Restore database backup
+   * @return {Promise<void>}
+   */
+  public async restoreDb (): Promise<void> {
+    const backUps = getBackUps()
+    const [_, oldest] = backUps
+
+    if (backUps.length < 2) {
+      throw new Error('Should be two backups to be able to restore')
+    }
+
+    // Check if back up block hash exist after reorg
+    const block = await this.eth.getBlock(oldest.block.hash).catch(() => false)
```

Comment:
I believe this is not what we want. You should fetch the block based on its block number that you have stored in the backup's name and then compare the block's hash with the hash you have stored in the backup's filename. If they match, great we have a valid backup if not, then you know that the reorg is bigger than the backup and then we are screwed 😅

Reply:
Ok, this is actually valid approach as well.